<div style="text-align:center">
<img src="img/intro_correspondre.png">
</div>


*** CREATION IN PROGRESS ***<br>
Notebook created by GITHUB ESTELLE DERRIEN

## Summary

15. <b> Modeling of a multiperiod production and storage problem </b>
    - 15.1 Taïwan university instance ( Coursera chpt 2.8 )
    - 15.2 Mathematical modeling
    - 15.3 Python Pulp solution ( Hard coded )
    - 15.4 Python Pulp solution ( Iterations )
    - 15.5 Columbia university instance
    - 15.6 Mathematical modeling
    - 15.7 Python Pulp solution and comparison
16. <b> Modeling of a multiperiod cash flow matching </b>
    - 16.1 Washington university instance 
    - 16.2 Mathematical modeling
    - 16.3 Python Pulp solution and comparison
    - 16.4 Columbia university instance
    - 16.5 Mathematical modeling
    - 16.6 Python Pulp solution and comparison
17. <b>Modeling of a Multiperiod Work Scheduling</b>
    - 17.1 Instance - Washington University.
    - 17.2 Mathematical modeling
    - 17.3 Python Pulp solution
18. <b>Modeling of an optimal employees time shift scheduling problem</b>
    - 18.1 Instance - Microsoft problem : "Using Solver to schedule your workforce".
    - 18.2 Mathematical modeling
    - 18.3 Python Pulp solution
19. <b>Modeling of a linear portfolio optimization problem</b>
    - 19.1 Instance by Excel easy
    - 19.2 Mathematical modeling
    - 19.3 Python Pulp solution
    - 19.4 Instance by Mastering python for finance p.96
    - 19.5 Mathematical modeling
    - 19.6 Python Pulp solution
    - 19.7 Instance by Kenneth A Barker
    - 19.8 Mathematical modeling
    - 19.9 Python Pulp solution
    - 19.10 Instance by Andrew Hershy
    - 19.11 Mathematical modeling
    - 16.3 Python Pulp solution
20. <b>Modeling of a facility location problem</b>
    - 20.1 Instance - Taïwan university problem
    - 20.2 Mathematical modeling
    - 20.3 Python Pulp solution and Chang Yaochen code analysis
21. <b> Modeling of a single machine scheduling problem</b>
    - 21.1 Instance
    - 21.2 Mathematical modeling
    - 21.3 Python Pulp solution
    - 21.4 Single machine, parallels ,flow shop and job shop differences.
22. <b> Modeling an uncapacited transport problem </b>
    - Presentation 
    - Mathematical modeling
    - Python Pulp or Cplex solution
23. <b> Modeling a capacited transport problem  </b>
    - Presentation of the Daskin problem (2013)
    - Mathematical modeling
    - Python Pulp or Cplex solution
        - Using hard values
        - Using iterations
24. <b> Modeling a bin packing problem  </b>
    - Presentation and differences between 2d and 3d
    - Mathematical modeling
    - Python Pulp or Cplex solution
25. <b> Modeling a traveling salesman problem  </b>
    - Presentation 
    - Mathematical modeling
    - Python Pulp or Cplex solution
26. <b> Modeling a shortest path problem  </b>
    - Presentation of https://blog.sommer-forst.de/2013/04/06/solving-the-minimum-cost-flow-problem-4-pulp/
    - https://transport-systems.imperial.ac.uk/tf/60008_21/n3_6_shortest_paths_pulp.html
    - Mathematical modeling
    - Python Pulp or Cplex solution
27. <b> Modeling a chemistry combinatorial optimization  </b>
    - Presentation of the Mac Graw Book problem BOILER/TURBO-GENERATOR SYSTEM
    - Mathematical modeling
    - Python Pulp or Cplex solution
28. <b> Modeling a petroleum combinatorial optimization </b>
    - Presentation and Kirchhoff constraints
    - Mathematical modeling
    - Python Pulp or Cplex solution
29. <b> Modeling of a multiple objectives problem</b>
    - Presentation of https://www.supplychaindataanalytics.com/multi-objective-linear-optimization-with-pulp-in-python/
    - Mathematical modeling
    - Python Pulp or Cplex solution
30. <b> Getting familiar with sensitivity analysis</b>
    - Displaying Reduced costs
    - Displaying The Shadow Price
    - Displaying Slacks variables
31. <b> Getting familiar with variables relaxations</b>

32. <b> Linearizations </b>
    - Linearization technics using the AIMMS modeling guide and Ling-Chieh Kung technics
    - Objective linearizations
    - Constraints linearizations
    - Examples using Python solvers


33. <b> Apendix </b> 
- Importing a data instance from an Excel file
- Importing a data instance from a Database

# <b>15. Modeling of a multiperiod production and storage problem </b>

<div style="text-align:center">
<img src="img/TopCable.jpg" width="500"/>
</div>

<b>15.1 Taïwan university Instance</b>

The manager predicts that we'll sell 100, 150, 200, and 170 on days 1, 2, 3, and 4 of the next week.

Production costs vary depending on the day of the week:
9th, 12th, 10th, and 12th on days 1, 2, 3, and 4 of the next week.

The storage cost is €1 per item stored, so if we keep it for 4 days, that's a cost of €4.

What is the best production plan to minimize costs and not have too much inventory?

<b>15.2 Math model</b><br>
Step 1 : Simplified/hard coded model :<br>
<img src="img/01. Formulation du L.P.png"></img>




<b>15.3 Python Pulp solution</b>
<div style="text-align:center">
<img src="img/small-pulp.png" width="100">
</div>

Hard coded version :

In [56]:
# -----------------------------------
# Import Pulp
# -----------------------------------
import pulp

# -----------------------------------
# Type of problem
# -----------------------------------
problem = pulp.LpProblem("multiperiod_lp",pulp.LpMinimize)

# -----------------------------------
# Decision variables
# -----------------------------------
# Items production per day 1,2,3,4
P1 = pulp.LpVariable('Prodution1', lowBound=0, cat='Integer')
P2 = pulp.LpVariable('Prodution2', lowBound=0, cat='Integer')
P3 = pulp.LpVariable('Prodution3', lowBound=0, cat='Integer')
P4 = pulp.LpVariable('Prodution4', lowBound=0, cat='Integer')


# Inventory per day 1,2,3,4
I1 = pulp.LpVariable('Inventory1', lowBound=0, cat='Integer')
I2 = pulp.LpVariable('Inventory2', lowBound=0, cat='Integer')
I3 = pulp.LpVariable('Inventory3', lowBound=0, cat='Integer')
I4 = pulp.LpVariable('Inventory4', lowBound=0, cat='Integer')

# -----------------------------------
# Math model
# -----------------------------------

# Objective function : minimizing production costs + inventory costs (1$ per item in inventory)
problem += (9*P1 + 12*P2 + 10*P3 + 12*P4) + (1*I1 + 1*I2 + 1*I3 + 1*I4) 

# Inventory balancing constraints
problem += P1 - 100 == I1
problem += I1 + P2 - 150 == I2
problem += I2 + P3 - 200 == I3
problem += I3 + P4 - 170 == I4

problem += P1 >= 0
problem += P2 >= 0 
problem += P3 >= 0 
problem += P4 >= 0 

# -----------------------------------
# Solution
# -----------------------------------
result = problem.solve()

# Printing solution
print("---First day---")
print("Production:" ,pulp.value(P1))
print("Inventory:" ,pulp.value(I1))


print("---2nd day---")
print("Production:" ,pulp.value(P2))
print("Inventory:" ,pulp.value(I2))

print("---3rd day---")
print("Production:" ,pulp.value(P3))
print("Inventory:" ,pulp.value(I3))

print("---4th day---")
print("Production:" ,pulp.value(P4))
print("Inventory:" ,pulp.value(I4))

print("Minimized production and storage cost = :" ,pulp.value(problem.objective))



---First day---
Production: 250.0
Inventory: 150.0
---2nd day---
Production: 0.0
Inventory: 0.0
---3rd day---
Production: 370.0
Inventory: 170.0
---4th day---
Production: 0.0
Inventory: 0.0
Minimized production and storage cost = : 6270.0


Our multi-period interval here is 4 days.

The solver calculates what needs to be produced and put into inventory in order to end up with no inventory on hand at the end of these 4 days, while still meeting demand.

On the first day, we produce 250 units, which meets the manager's forecast of selling 100 units on the first day, and we keep 150 units in inventory for the second day.

On the third day, we produce 370 units, still with the aim of meeting the manager's forecast for the following 2 days, and we keep 170 units in inventory.

On the fourth day, we have minimized our inventory and our carrying costs, and we have nothing left.

<b>15.4 Python Pulp solution ( Using instance and iterations)</b>

<b>15.4.1 Math model </b>
Step 2 : Compact model :
$$
Min (C) = \sum_{t=1}^{4} (c_{t}.x_{t} + sc_{t}.y_{t})
$$
$$
s.t
$$
$$
  y_{t-1} + x_{t} - d_{t} = y_{t} \thinspace ∀ \thinspace t \in (1,...,4)
$$
$$
y_{0} = 0
$$

$$
where
$$

$$
 c_{t} = production \thinspace cost
$$
$$
 sc_{t} = storage \thinspace cost
$$
$$
 d_{t} = demand
$$


<div style="text-align:center">
<img src="img/small-pulp.png" width="100">
</div>

In [57]:
# -----------------------------------
# Import Pulp
# -----------------------------------
from pulp import *

# -----------------------------------
# Type of problem
# -----------------------------------
model = LpProblem("multiperiod_lp",LpMinimize)

# -----------------------------------
# Instance
# -----------------------------------
# Define the production and storage costs for days 1, 2, 3, and 4; in quarters, it's the number of days, i.e., 4.
# Define production cost, inventory cost, and demand.
# Define the production costs per period, the inventory cost, and the demand reported by the sales department.
quaters = list(range(4))
c       =[9,12,10,12] # production cost
sc      =[1,1,1,1]
d       =[100,150,200,170]

# -----------------------------------
# Decision variables
# -----------------------------------
x = LpVariable.dicts('quater_prod_', quaters,lowBound=0, cat='Continuous')
y = LpVariable.dicts('quater_inventory_', quaters,lowBound=0, cat='Continuous')

# -----------------------------------
# Math model
# -----------------------------------
# Objective
model += lpSum([c[t]*x[t] for t in quaters]) + lpSum([sc[t]*y[t] for t in quaters])

# Constraints

# Inventory-balance constraints
model.addConstraint(x[0] - y[0] == d[0]) # (day 1)

for t in quaters[1:]:
    model.addConstraint(x[t] - y[t] + y[t-1] == d[t]) # (days 2, 3, 4) 

# Production-capacity constraints - optional
for i in quaters:
    model.addConstraint(x[i]<=3000)

# -----------------------------------
# Solution
# -----------------------------------
model.solve()

for v in model.variables():
    print(v.name, "=", v.varValue)
    
print("Minimized production and storage cost =  ", value(model.objective))

quater_inventory__0 = 150.0
quater_inventory__1 = 0.0
quater_inventory__2 = 170.0
quater_inventory__3 = 0.0
quater_prod__0 = 250.0
quater_prod__1 = 0.0
quater_prod__2 = 370.0
quater_prod__3 = 0.0
Minimized production and storage cost =   6270.0


<b>15.5 Columbia university Instance</b>

<b>15.6 Instance</b><br>

National Steel Corporation (NSC) produces a special-purpose steel
used in the aircraft and aerospace industries. The sales department
has received orders for the next four months...

Read : http://www.columbia.edu/itc/sipa/U6033/client_edit/lectures/lec4.pdf




NSC can meet demand by producing steel, drawing on its
inventory, or a combination of these.Inventory at the start of
January is zero.Production costs are expected to increase in February and March.
Production and stock costs are:

<table>
  <tr>
    <th></th>
    <th>January</th>
     <th>February</th>
      <th>March</th>
       <th>April</th>
  </tr>
  <tr>
  <td>Production cost</td>
    <td>3000</td>
    <td>3300</td>
    <td>3600</td>
    <td>3600</td>
  </tr>
    <tr>
  <td>Inventory cost</td>
    <td>250</td>
    <td>250</td>
    <td>250</td>
    <td>250</td>
  </tr>
</table>

Production costs are dollars per tonne. 
Inventory costs are in dollars per tonne and per month. 

For example, 1 ton in stock for 1 month costs 250 dollars;for 2
months it costs $ 500.

NSC can produce a maximum of 3,000 tonnes of steel per month. 
What production plan does it respond to demand at the minimum cost?

<b>15.7 Maths model</b><br>
Compact model :
$$
Min (C) = \sum_{t=1}^{4} (c_{t}.x_{t} + sc_{t}.y_{t})
$$
$$
s.t
$$
$$
  y_{t-1} + x_{t} - d_{t} = y_{t} \thinspace ∀ \thinspace t \in (1,...,4)
$$
$$
y_{0} = 0
$$

$$
where
$$

$$
 c_{t} = production \thinspace cost
$$
$$
 sc_{t} = storage \thinspace cost
$$
$$
 d_{t} = demand
$$


<b>15.8 Python Pulp solution</b>
<div style="text-align:center">
<img src="img/small-pulp.png" width="100">
</div>

In [58]:
# -----------------------------------
# Import Pulp
# -----------------------------------
from pulp import *

# -----------------------------------
# Type of problem
# -----------------------------------
model = LpProblem("multiperiod_production_minimize_cost",LpMinimize)

# -----------------------------------
# Instance
# -----------------------------------
quaters =   list(range(4))
c       =   [3000, 3300, 3600, 3600]
sc      =   [250, 250, 250, 250]
d       =   [2300, 2000, 3100, 3000]

# -----------------------------------
# Decision variables
# -----------------------------------
x = LpVariable.dicts('quater_prod_', quaters,lowBound=0, cat='Continuous')
y = LpVariable.dicts('quater_inventory_', quaters,lowBound=0, cat='Continuous')

# -----------------------------------
# Math model
# -----------------------------------
# Objective
model += lpSum([c[t]*x[t] for t in quaters]) + lpSum([sc[t]*y[t] for t in quaters])

# Constraints

# Inventory-balance constraints
model.addConstraint(x[0] - y[0] == d[0]) # (day 1)

for t in quaters[1:]:
    model.addConstraint(x[t] - y[t] + y[t-1] == d[t]) # (days 2, 3, 4) 

# Production-capacity constraints - optional
for i in quaters:
    model.addConstraint(x[i]<=3000)

# -----------------------------------
# Solution
# -----------------------------------
model.solve()

for v in model.variables():
    print(v.name, "=", v.varValue)
    
print("Minimized production and storage cost =  ", value(model.objective))

quater_inventory__0 = 700.0
quater_inventory__1 = 1700.0
quater_inventory__2 = 0.0
quater_inventory__3 = 0.0
quater_prod__0 = 3000.0
quater_prod__1 = 3000.0
quater_prod__2 = 1400.0
quater_prod__3 = 3000.0
Minimized production and storage cost =   35340000.0


We see that the python pulp solution is the same than the Columbia university website one, looking good, bro.

<div style="text-align:center">
<img src="img/thumbs-up.jpg" width="100">
</div>

# <b> 16.  Modeling of a multiperiod cash flow matching </b>

<div style="text-align:center">
<img src="img/cfm.webp" width="500"/>
</div>

<b>16.1 Washington university Instance</b>




<b>16.2 Math model</b><br>
Comin' soon

<b>16.3 Python Pulp solution</b>
<div style="text-align:center">
<img src="img/small-pulp.png" width="100">
</div>

In [59]:
# Import
from pulp import *

# 1. Set problem : maximize cash on hand at time 3
problem = LpProblem("cashFlow_matching",LpMaximize)

# Define variables

# Finco must decide how much money should be placed in each investment (including
# money market funds). Thus, we define the following decision variables:

# Dollars invested in investments 
A = pulp.LpVariable('A', lowBound=0, cat='Integer')
B = pulp.LpVariable('B', lowBound=0, cat='Integer')
C = pulp.LpVariable('C', lowBound=0, cat='Integer')
D = pulp.LpVariable('D', lowBound=0, cat='Integer')
E = pulp.LpVariable('E', lowBound=0, cat='Integer')

# Dollars invested in in money market funds at time i
S0 = pulp.LpVariable('S0', lowBound=0, cat='Integer')
S1 = pulp.LpVariable('S1', lowBound=0, cat='Integer')
S2 = pulp.LpVariable('S2', lowBound=0, cat='Integer')

# Finco wants to maximize cash on hand at time 3. At time 3, Finco’s cash on hand will be
# the sum of all cash inflows at time 3. From the description of investments A–E and the
# fact that from time 2 to time 3, S2 will increase to 1.08 * S2,

# Finco souhaite maximiser ses liquidités à l'instant 3. 
# a l'instant 3, les liquidités de Finco seront de
# la somme de toutes les entrées de trésorerie au moment 3. 
# À partir de la description des investissements A–E et du
# fait que du temps 2 au temps 3, S2 augmentera jusqu'à 1,08 * S2,

# Objective : maximize cash on hand at time 3
problem += B + 1.9 * D + 1.5 * E + 1.08 * S2

# Constraints
# investment starting point ( B and E are not available at time 0)
problem += A  + C + D + S0 == 100000

# Time 1
problem += 0.5 * A + 1.2 * C + 1.08 * S0 == B + S1

# Time 2
problem += A + 0.5 * B + 1.08 * S1 == E + S2

# Max invest
problem += A <= 75000
problem += B <= 75000
problem += C <= 75000
problem += D <= 75000
problem += E <= 75000

# Non negativity
problem +=  A >= 0
problem +=  B >= 0
problem +=  C >= 0
problem +=  D >= 0
problem +=  E >= 0
problem +=  S0 >= 0
problem +=  S1 >= 0
problem +=  S2 >= 0




# The problem is solved using PuLP's choice of Solver
problem.solve()

# Print the variables optimized value
for v in problem.variables():
    print(v.name, "=", v.varValue)
    
# The optimised objective function value is printed to the screen
print("Maximized portfolio at time 3 = ", value(problem.objective), " in $ millions")

# We find the optimal solution to be z = 218,500, A  = 60,000, B = 30,000, D  = 40,000,
# E = 75,000, C = S0 = S1 = S2 = 0. Thus, Finco should not invest in money market
# funds. At time 0, Finco should invest $60,000 in A and $40,000 in D. Then, at time 1, the
# $30,000 cash inflow from A should be invested in B. Finally, at time 2, the $60,000 cash
# inflow from A and the $15,000 cash inflow from B should be invested in E. At time 3,
# Finco’s $100,000 will have grown to $218,500

A = 60000.0
B = 30000.0
C = 0.0
D = 40000.0
E = 75000.0
S0 = 0.0
S1 = 0.0
S2 = 0.0
Maximized portfolio at time 3 =  218500.0  in $ millions
